X9.146/gen_dilithium2_dual_keysig_cert.c

```diff
@@ -0,0 +1,379 @@
+#include <wolfssl/wolfcrypt/dilithium.h>
+#include <wolfssl/wolfcrypt/asn_public.h>
+#include <wolfssl/wolfcrypt/asn.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/logging.h>
+
+#if defined(WOLFSSL_X9_146) && defined(HAVE_LIBOQS)
+
+#define LARGE_TEMP_SZ 9216
+
+#if defined(GEN_ROOT_CERT) && defined(GEN_SERVER_CERT)
+    #error "Please only generate a root OR server certificate."
+#endif
+
+#define SUBJECT_COUNTRY "US"
+#define SUBJECT_STATE "MT"
+#define SUBJECT_LOCALITY "YourCity"
+#define SUBJECT_ORG "YourOrgName"
+#define SUBJECT_UNIT "YourUnitName"
+#define SUBJECT_COMMONNAME "www.YourDomain.com"
+
+#ifdef GEN_ROOT_CERT
+#define SUBJECT_EMAIL "pq-root@YourDomain.com"
+#else
+#define SUBJECT_EMAIL "pq-server@YourDomain.com"
+#endif
+
+static int do_certgen(int argc, char** argv)
+{
+    int ret = 0;
+
+    char caKeyFile[] = "./ca-key.der";
+#ifdef GEN_ROOT_CERT
+    char newCertOutput[] = "./ca-cert-pq.der";
+    char sapkiFile[] = "../certs/dilithium_level2_ca_pubkey.der";
+    char altPrivFile[] = "../certs/dilithium_level2_ca_key.der";
+#else
+    char caCert[] = "./ca-cert-pq.der";
+    char newCertOutput[] = "./server-cert-pq.der";
+    char sapkiFile[] = "../certs/dilithium_level2_server_pubkey.der";
+    char altPrivFile[] = "../certs/dilithium_level2_server_key.der";
+    char serverKeyFile[] = "./server-key.der";
+#endif
+    FILE* file;
+    Cert newCert;
+    DecodedCert preTBS;
+
+#ifndef GEN_ROOT_CERT
+    byte caCertBuf[LARGE_TEMP_SZ];
+    int caCertSz = LARGE_TEMP_SZ;
+    byte serverKeyBuf[LARGE_TEMP_SZ];
+    int serverKeySz = LARGE_TEMP_SZ;
+#endif /* !GEN_ROOT_CERT */
+
+    byte caKeyBuf[LARGE_TEMP_SZ];
+    int caKeySz = LARGE_TEMP_SZ;
+    byte sapkiBuf[LARGE_TEMP_SZ];
+    int sapkiSz = LARGE_TEMP_SZ;
+    byte altPrivBuf[LARGE_TEMP_SZ];
+    int altPrivSz = LARGE_TEMP_SZ;
+    byte altSigAlgBuf[LARGE_TEMP_SZ];
+    int altSigAlgSz = LARGE_TEMP_SZ;
+    byte scratchBuf[LARGE_TEMP_SZ];
+    int scratchSz = LARGE_TEMP_SZ;
+    byte preTbsBuf[LARGE_TEMP_SZ];
+    int preTbsSz = LARGE_TEMP_SZ;
+    byte altSigValBuf[LARGE_TEMP_SZ];
+    int altSigValSz = LARGE_TEMP_SZ;
+    byte outBuf[LARGE_TEMP_SZ];
+    int outSz = LARGE_TEMP_SZ;
+
+    /* The are for MakeCert and SignCert. */
+    WC_RNG rng;
+    int initRng = 0;
+
+    RsaKey caKey;
+    int initCaKey = 0;
+#ifndef GEN_ROOT_CERT
+    RsaKey serverKey;
+    int initServerKey = 0;
+#endif /* !GEN_ROOT_CERT */
+    int initPreTBS = 0;
+    dilithium_key altCaKey;
+    word32 idx = 0;
+
+#if 0
+    wolfSSL_Debugging_ON();
+#endif
+
+    ret = wc_InitRng(&rng);
+    if (ret != 0) goto exit;
+    initRng = 1;
+
+#ifndef GEN_ROOT_CERT
+    /* Open the CA der formatted certificate. if we are generating the server
+     * certificate. We need to get it's subject line to use in the new cert
+     * we're creating as the "Issuer" line */
+    printf("Loading CA certificate\n");
+    XMEMSET(caCertBuf, 0, caCertSz);
+    file = fopen(caCert, "rb");
```

Comment:
Extract out opening a file and reading into a separate function.

Reply:
Yeah, we do that over and over again so I can do that.